activesupport/lib/active_support/cache/redis_cache_store.rb

```diff
@@ -445,7 +445,7 @@ def change_counter(key, amount, options)
               end.first
             else
               count = c.incrby(key, amount)
-              if count != amount && options[:expires_in] && c.ttl(key) < 0
```

Comment:
I have a cold so can't think straight enough to be confident. But I think the condition should have been `count == amount`.

The reasoning being that if the counter already existing you don't need to set its TTL, it should already have one.

Reply:
I believe that condition is covered by `c.ttl(key) <0`, as the TTL will be negative if it hasn't been set or if it has expired.

> In Redis 2.6 or older the command returns -1 if the key does not exist or if the key exist but has no associated expire.
> 
> Starting with Redis 2.8 the return value in case of error changed:
> 
> The command returns -2 if the key does not exist.
> The command returns -1 if the key exists but has no associated expire.

_source: https://redis.io/commands/ttl/_

Reply:
The point is that `client.ttl` is a full roundtrip to the Redis server, so it might take almost 1ms.

That's why this extra `count == amount` condition exist, to avoid doing 2 full roundtrip to the server instead of a singe one when incrementing.

Reply:
> the trade off being that using `count == amount` introduces new and divergent behavior between Redis 7+ and Redis <7.

How so? Again I'm not thinking straight today, but I think it checks out.

If the key didn't exist in Redis, then the value returned by `incrby` will match `amount`:

```
127.0.0.1:6379> INCRBY counter 2
(integer) 2
127.0.0.1:6379> INCRBY counter2 -5
(integer) -5
```

I think this is a good enough proxy to consider setting the expiry. I'm even tempted to not check the `TTL` at all.

Reply:
Oh yea! for non-existent keys that works well.

The tests that are failing are for keys that do _exist_ and _do not have a TTL_ and are expect to have a TTL after `#increment` or `#decrement` is called.

```ruby
# key exist but not have expire
redis_backend { |r| r.set "#{@namespace}:dar", 10 }
@cache.increment "dar", 1, expires_in: 60
  redis_backend do |r|
  assert r.ttl("#{@namespace}:dar") > 0
end
```

Diverging Behavior: 
- With Redis 7, we'll set the TTL only if it doesn't exist because of the `NX` operator in the pipeline.
- Redis <7, `count == 11` and `amount == 1`.

I'm not sure if this is an expected usecase, and I'm not necessarily advocating for making the additional TTL call 😅 I'm mostly interested in removing/correcting the tests to match the behavior.

Reply:
I think I'm OK for this small discrepancy in such a corner case if it means noticeably better performance for the vast majority of use cases.

Reply:
Another possibility is to pipeline the `INCRBY` and `TTL` together.

Reply:
> Another possibility is to pipeline the `INCRBY` and `TTL` together.

I thought so too, however `NX` for `EXPIRE` was only introduced in Redis 7.0.0, so we lose the set-ttl-if-ttl-not-exist functionality. I.e., we would re-set the TTL on every `#increment`/`#decrement` call.

Reply:
Oh? TIL you can put a condition in `pipelined`!

Random example:

```ruby
key = "foo"
=> "foo"

c.pipelined do |pipeline|
  pipeline.incrby(key, 1)
  pipeline.call(:expire, key, 10) if c.ttl(key) < 0
end
=> [1, 1]

c.ttl key
=> 8

c.ttl key
=> 5

c.pipelined do |pipeline|
  pipeline.incrby(key, 1)
  pipeline.call(:expire, key, 10) if c.ttl(key) < 0
end
=> [2]

c.ttl key
=> 1
```

I'm going to check the network just to make sure this is only making the calls I think it's making.

Reply:
> Oh? TIL you can put a condition in `pipelined`!

No you can't. This will evaludate `TTL` first, and the pipeline after.